ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+	} else {
+		task.groups = append(task.groups, wg)
+		task.groupIndices = append(task.groupIndices, wg.balancesCount)
+	}
+	wg.balancesCount++
+}
+
+func loadAccountsAddResourceTask(addr *basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType, wg *groupTask, resourceTasks map[accountCreatableKey]*preloaderTask, queue *preloaderTaskQueue) {
+	if cidx == 0 {
+		return
+	}
+	key := accountCreatableKey{
+		cidx: cidx,
+	}
+	if addr != nil {
+		key.address = *addr
+	}
+	if task, have := resourceTasks[key]; !have {
+		task := &preloaderTask{
+			address:        addr,
+			groups:         make([]*groupTask, 1, 4),
+			groupIndices:   make([]int, 1, 4),
+			creatableIndex: cidx,
+			creatableType:  ctype,
+		}
+		task.groups[0] = wg
+		task.groupIndices[0] = wg.resourcesCount
+
+		resourceTasks[key] = task
+		queue.enqueue(task)
+	} else {
+		task.groups = append(task.groups, wg)
+		task.groupIndices = append(task.groupIndices, wg.resourcesCount)
+	}
+	wg.resourcesCount++
+}
+
+func (p *accountPrefetcher) prefetch(ctx context.Context) {
+	defer close(p.outChan)
+	accountTasks := make(map[basics.Address]*preloaderTask)
+	resourceTasks := make(map[accountCreatableKey]*preloaderTask)
+
+	var maxTxnGroupEntries int
+	if p.consensusParams.Application {
+		// the extra two are for the sender account data, plus the application global state
+		maxTxnGroupEntries = p.consensusParams.MaxTxGroupSize * (2 + p.consensusParams.MaxAppTxnAccounts + p.consensusParams.MaxAppTxnForeignApps + p.consensusParams.MaxAppTxnForeignAssets)
+	} else {
+		// 8 is the number of resources+account used in the AssetTransferTx, which is the largest one.
+		maxTxnGroupEntries = p.consensusParams.MaxTxGroupSize * 8
+	}
+
+	tasksQueue := allocPreloaderQueue(len(p.groups), maxTxnGroupEntries)
+
+	// totalBalances counts the total number of balances over all the transaction groups
+	totalBalances := 0
+	totalResources := 0
+
+	groupsReady := make([]groupTask, len(p.groups))
+
+	// Add fee sink to the first group
+	if len(p.groups) > 0 {
+		// the feeSinkAddr is known to be non-empty
+		feeSinkPreloader := &preloaderTask{
+			address:      &p.feeSinkAddr,
+			groups:       []*groupTask{&groupsReady[0]},
+			groupIndices: []int{0},
+		}
+		groupsReady[0].balancesCount = 1
+		accountTasks[p.feeSinkAddr] = feeSinkPreloader
+		tasksQueue.enqueue(feeSinkPreloader)
+	}
+
+	// iterate over the transaction groups and add all their account addresses to the list
+	queue := &tasksQueue
+	for i := range p.groups {
+		task := &groupsReady[i]
+		for j := range p.groups[i] {
+			stxn := &p.groups[i][j]
+			switch stxn.Txn.Type {
+			case protocol.PaymentTx:
+				loadAccountsAddAccountTask(&stxn.Txn.Receiver, task, accountTasks, queue)
+				loadAccountsAddAccountTask(&stxn.Txn.CloseRemainderTo, task, accountTasks, queue)
+			case protocol.AssetConfigTx:
+				loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.ConfigAsset), basics.AssetCreatable, task, resourceTasks, queue)
```

Comment:
why addr is nil here? I'd expect to preload sender's asset params here and ignore for stxn.Txn.ConfigAsset == 0

Reply:
look at AssetConfig  in apply/asset.go :
`if cc.ConfigAsset != 0`, then it would load the parameters for the `cc.ConfigAsset` that asset record doesn't necessarily related to the sender ( i.e. it could be the manager as well ). By passing nil, we're looking for the asset creator so we can load the asset parameters.